mainframer.sh

```diff
@@ -36,83 +36,31 @@ if [ -z "$BUILD_COMMAND" ]; then
 	exit 1
 fi
 
-# Create build folder in case if it does not exist.
-mkdir -p "$PROJECT_DIR"/build
-
-# Remove previous archives of the project.
-rm -f "$PROJECT_DIR"/build/project_for_remote_build.tar "$PROJECT_DIR"/build/remotely_built_project.tar
-
-# Archiver.
-
-DETECT_ARCHIVER="if type 'pigz' > /dev/null; then echo 'pigz'; else echo 'gzip'; fi"
-LOCAL_ARCHIVER=`eval $DETECT_ARCHIVER`
-
-# Archive project.
 pushd "$PROJECT_DIR"
-LOCAL_ARCHIVE_COMMAND="tar \
--c \
---exclude='build/project_for_remote_build.tar' \
---exclude='local.properties' \
+# Sync project to remote machine.
+rsync -a --delete --compress-level=$LOCAL_COMPRESS_LEVEL \
 --exclude='.gradle' \
 --exclude='.idea' \
---exclude='.git' \
+--exclude='**/.git/' \
 --exclude='artifacts' \
 --exclude='captures' \
---exclude='build' \
---exclude='*/build' \
-."
-
-if [ $LOCAL_GZIP_LEVEL = "0" ]; then
-	LOCAL_ARCHIVE_COMMAND+=" > build/project_for_remote_build.tar"
-	REMOTE_UNARCHIVE_COMMAND="tar -xf project_for_remote_build.tar -C $PROJECT_DIR_NAME"
-else
-	LOCAL_ARCHIVE_COMMAND+=" | $LOCAL_ARCHIVER -$LOCAL_GZIP_LEVEL > build/project_for_remote_build.tar"
-	REMOTE_UNARCHIVE_COMMAND="\$REMOTE_ARCHIVER -d < project_for_remote_build.tar | tar -xf - -C $PROJECT_DIR_NAME"
-fi
-
-eval $LOCAL_ARCHIVE_COMMAND
-popd
+--exclude='**/build' \
+-e "ssh" ./ "$REMOTE_BUILD_MACHINE:~/$PROJECT_DIR_NAME"
 
-# Prepare remote archive and local unarchive commands.
-REMOTE_ARCHIVE_COMMAND="tar \
--c \
---exclude='kotlin' \
---exclude='tmp' \
-build/ */build"
-
-if [ $REMOTE_GZIP_LEVEL = "0" ]; then
-	REMOTE_ARCHIVE_COMMAND+=" > remotely_built_project.tar"
-	LOCAL_UNARCHIVE_COMMAND="tar -xf build/remotely_built_project.tar -C ./"
-else
-	REMOTE_ARCHIVE_COMMAND+=" | \$REMOTE_ARCHIVER -$REMOTE_GZIP_LEVEL > remotely_built_project.tar"
-	LOCAL_UNARCHIVE_COMMAND="$LOCAL_ARCHIVER -d < build/remotely_built_project.tar | tar -xf - -C ./"
-fi
-
-# Transfer archive to remote machine.
-scp "$PROJECT_DIR/build/project_for_remote_build.tar" $REMOTE_BUILD_MACHINE:~/
-
-# Build project on a remote machine and then archive it.
+# Build project on a remote machine.
 ssh $REMOTE_BUILD_MACHINE \
 "set -xe && \
-export REMOTE_ARCHIVER=\`eval \"$DETECT_ARCHIVER\"\` && \
-cd ~ && \
-mkdir -p $PROJECT_DIR_NAME && \
-rm -rf $PROJECT_DIR_NAME/build/remotely_built_project.tar $PROJECT_DIR_NAME/*/src && \
-$REMOTE_UNARCHIVE_COMMAND && \
-cd $PROJECT_DIR_NAME && \
-$BUILD_COMMAND && \
-$REMOTE_ARCHIVE_COMMAND"
-
-# Clean local build dirs.
-rm -rf "$PROJECT_DIR"/build "$PROJECT_DIR"/*/build
-mkdir -p "$PROJECT_DIR/build/"
+cd ~/$PROJECT_DIR_NAME/ && \
+$BUILD_COMMAND"
 
-# Copy build results from remote machine to local.
-scp "$REMOTE_BUILD_MACHINE":~/"$PROJECT_DIR_NAME"/remotely_built_project.tar "$PROJECT_DIR/build/"
-
-# Unarchive build results.
-pushd "$PROJECT_DIR"
-eval "$LOCAL_UNARCHIVE_COMMAND"
+# Sync project back to local machine.
+rsync -a --delete --compress-level=$REMOTE_COMPRESS_LEVEL \
```

Comment:
Don’t know about @artem-zinnatullin, but I would prefer full parameter names, i. e. `--archive` instead of `-a`.